### PR TITLE
fast5_archive datatype

### PR DESCRIPTION
**Pull request by jvolkening**
FAST5 files are HDF5-based sequence containers produced by Oxford Nanopore sequencers. Typically one file is produced per read, so hundreds of thousands to millions of files can be produced in a run. To make management easier, sets of related FAST5 files can be compressed and distributed as tarballs.

I'm wrapping some nanopore-related software and need to support FAST5 input/output. This PR adds a `fast5_archive` datatype (inspired by the `searchgui_archive` type from `GalaxyP`) which is simply a tarball containing FAST5 files in any directory structure which can be sniffed but is marked to not be automatically decompressed. I considered using collections of individual FAST5 files but this approach seemed easier to manage and easier on the backend database.

### CODE REVIEW

**Comment by jvolkening** [comment]
Incidentally, it would be nice if the proper file extension (`tar`, `tar.gz`, `tar.bz2` was appended on download, rather than the `fast5_archive` extension which is meaningless outside Galaxy. Is this possible? It can't be hard-coded since any of the above are accepted. I could check the magic bits to determine the correct download extension dynamically if I knew where to do this in the code.

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -1487,6 +1489,54 @@ def sniff(self, filename):
...
+                temptar.close()
+        except Exception as e:
+            log.warning( '%s, set_meta Exception: %s', self, e )
+
+    def sniff( self, filename ):
+        try:
+            if filename and tarfile.is_tarfile( filename ):
+                temptar = tarfile.open( filename, 'r' )
+                contents = temptar.getnames()
+                r = re.compile(".+\.fast5$")
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: I think `return any((c for c in contents if c.endswith('.fast5')))` would be preferable here, or if we're talking millions of files then
```python
with open('filename', 'r') as temptar:
    contents = temptar.getnames()
    for content in contents:
        if content.endswith('.fast5'):
            return True
```
might have better performance than `filter` on the whole list.

File: lib/galaxy/datatypes/binary.py [hunk 2]
```diff
@@ -1487,6 +1489,54 @@ def sniff(self, filename):
...
+                     readonly=True, visible=True, no_value=None )
+    file_ext = "fast5_archive"
+
+    def set_meta( self, dataset, overwrite=True, **kwd ):
+        super( Fast5Archive, self ).set_meta( dataset, overwrite=overwrite, **kwd )
+        try:
+            if dataset and tarfile.is_tarfile( dataset.file_name ):
+                temptar = tarfile.open( dataset.file_name, 'r' )
+                contents = temptar.getnames()
+                r = re.compile(".+\.fast5$")
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: Can we replace the regex with `.endswith('.fast5')`?
Also `filter` returns an iterator on python3, so I guess the most efficient way to get the count would be `sum(1 for content in contents if content.endswith('.fast5'))`

**Review by mvdbeek — COMMENTED** [review]
Looks nice, I like the approach.

**Comment by mvdbeek** [comment]
> , it would be nice if the proper file extension (tar, tar.gz, tar.bz2 was appended on download,

The tar members may be compressed in either of these formats? One option would be to have subclasses for each of the compression types that append the extension. I don't think we have a dynamic way of setting the extension at the moment.

**Comment by mvdbeek** [comment]
(also travis complains about the whitespace around parens )

**Comment by mvdbeek** [comment]
@galaxybot test this

**Comment by jvolkening** [comment]
@mvdbeek Thanks for the comments. I will implement your suggestions, fix the formatting, and re-test.

**Comment by jvolkening** [comment]
@mvdbeek I implemented your optimizations. Actually, I took the sniffing one step further and rather than slurping all filenames I just iterate over them and return at the first `.fast5` file that is seen, which is very fast. There are still possible performance issues if anyone adds another datatype involving tarballs of many files which is sniffed first, since it will look through the whole tarball and never find any `.fast5` files, but that is a future issue to be addressed.

**Comment by jvolkening** [comment]
> One option would be to have subclasses for each of the compression types that append the extension

I tried to implement this -- please review the changes. However, it does not seem to be working as expected. If I manually call the `sniff()` method of the various subclasses on different tarballs from within a shell, they return the expected values. However, if I fire up Galaxy and try to upload a `fast5.tar.gz` file, for instance, it is detected as a plain `fast5.tar` and given the wrong extension. I suspect that this has to do with the order the sniffers are called in, but I tried to modify `datatypes_conf.xml` as per the documentation (also in this PR).

**Comment by nsoranzo** [comment]
@jvolkening Does the archive potentially contain non-fast5 files? If not, you can just return `False` immediately.

**Comment by mvdbeek** [comment]
Could you include a small example of each dataype in https://github.com/galaxyproject/galaxy/tree/dev/lib/galaxy/datatypes/test, then we could add a doctest as in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/sequence.py#L603

**Comment by mvdbeek** [comment]
> However, if I fire up Galaxy and try to upload a fast5.tar.gz file, for instance, it is detected as a plain fast5.tar

Is that with datatype set to `auto` ? I'm afraid if you don't set the compressed variant of this we're just going to unpack, but I'm not sure that's the problem.

**Comment by jvolkening** [comment]
> Does the archive potentially contain non-fast5 files? If not, you can just return False immediately.

It can contain contain any number and depth of directories, which I believe would be listed by python's `tarfile`.

**Comment by jvolkening** [comment]
> Is that with datatype set to auto ? I'm afraid if you don't set the compressed variant of this we're just going to unpack, but I'm not sure that's the problem.

Yes, this behavior is with auto-detect. The subclasses just check the first few bytes for gzip or bzip2 magic numbers and then pass the rest of the sniffing off to the parental class. If the file is being decompressed before the sniffer sees it, that would explain the problem.

**Comment by jvolkening** [comment]
> Could you include a small example of each dataype

Done.

**Comment by jvolkening** [comment]
If the subclassing is too much of a hassle, I will just go back to the previous behavior, which was working fine with my tools. The only real issue here is the naming of downloaded files, which is a fairly minor annoyance.

**Comment by jvolkening** [comment]
> Does the archive potentially contain non-fast5 files? If not, you can just return False immediately.

@nsoranzo Here is my solution stemming from your comment:
```
for c in temptar:
    if not c.isfile():
        continue
    if c.name.endswith('.fast5'):
        return True
    else:
        return False
```

I believe this should short-circuit quickly in all situations.

**Comment by nsoranzo** [comment]
@jvolkening That looks good to me.

**Comment by jvolkening** [comment]
> if I fire up Galaxy and try to upload a fast5.tar.gz file, for instance, it is detected as a plain fast5.tar and given the wrong extension. I suspect that this has to do with the order the sniffers are called in, but I tried to modify datatypes_conf.xml as per the documentation (also in this PR).

I believe this is fixed -- it appears the sniff order is dependent on the order in which `register_sniffable_binary_format()` is called, rather than the order given in `datatypes_conf.xml`

After further testing and cleaning up the lint error from Travis CI, I will push a (hopefully final) commit.

**Comment by jvolkening** [comment]
> we could add a doctest as in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/sequence.py#L603

@mvdbeek Is there a way to run these tests locally before I push another commit?

Edit: nevermind, I figured out `nosetests --with-doctest`

**Comment by mvdbeek** [comment]
> Edit: nevermind, I figured out nosetests --with-doctest

Right, you can also use `tox -e p27-unit`, which will setup a new clean environment.

**Comment by mvdbeek** [comment]
@galaxybot test this

**Review by mvdbeek — APPROVED** [review]

**Comment by mvdbeek** [comment]
@nsoranzo is this OK for you as well ?

**Comment by nsoranzo** [comment]
Looks good to me.

**Comment by mvdbeek** [comment]
Thank you very much @jvolkening, this is wonderful. Looking forward to nanopore tools in galaxy!

**Pull request merged**